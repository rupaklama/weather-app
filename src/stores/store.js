import { createContext, useContext } from "react";
import WeatherStore from "./weatherStore";

export const store = {
  weatherStore: new WeatherStore(),
};

export const StoreContext = createContext(store);

export const useStore = () => {
  return useContext(StoreContext);
};
