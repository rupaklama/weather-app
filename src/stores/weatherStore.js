/* eslint-disable prettier/prettier */
import { makeAutoObservable, runInAction } from "mobx";

export default class WeatherStore {
  weather = null;

  initialLoading = false;

  error = null;

  constructor() {
    makeAutoObservable(this);
  }

  fetchWeather = async query => {
    this.initialLoading = true;

    try {
      const response = await fetch(
        `https://api.openweathermap.org/data/2.5/weather?q=${query}&APPID=dd9dbc16899659de44ea30ebc32456a9&units=imperial`
      );

      const data = await response.json();

      if (response.status === 200) {
        runInAction(() => {
          this.weather = data;
          this.initialLoading = false;
        });
      }

      if (response.status !== 200) {
        runInAction(() => {
          this.initialLoading = false;
          this.error = data;
        });

        // setTimeout(() => {
        //   window.location.reload();
        // }, 1500);
      }
    } catch (err) {
      console.log(err);

      runInAction(() => {
        this.initialLoading = false;
        this.error = err;
      });
    }
  };
}
