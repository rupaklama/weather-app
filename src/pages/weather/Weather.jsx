/* eslint-disable consistent-return */
/* eslint-disable react/no-array-index-key */
/* eslint-disable prettier/prettier */
import { observer } from "mobx-react-lite";

import React, { useEffect, useState } from "react";
import RotateLoader from "react-spinners/RotateLoader";

import Button from "../../components/button/Button";
import Input from "../../components/input/Input";
import styles from "./Weather.module.css";

import { useStore } from "../../stores/store";

const override = {
  display: "block",
  margin: "0 auto",
};

const Weather = observer(() => {
  const { weatherStore } = useStore();
  const { initialLoading, error, weather } = weatherStore;

  const [userCoordinates, setUserCoordinates] = useState({ lon: "", lat: "" });

  const [userCity, setUserCity] = useState("");
  const [searchText, setSearchText] = useState("");
  const [searchQueries, setSearchQueries] = useState([]);

  useEffect(() => {
    window.navigator.geolocation?.getCurrentPosition(
      position => setUserCoordinates({ lon: position.coords.longitude, lat: position.coords.latitude }),

      err => console.log(err)
    );

    if (userCoordinates.lat && userCoordinates.lon) {
      const fetchUserCity = () =>
        fetch(
          `https://api.openweathermap.org/geo/1.0/reverse?lat=${userCoordinates.lat}&lon=${userCoordinates.lon}&appid=dd9dbc16899659de44ea30ebc32456a9`
        )
          .then(res => res.json())
          .then(data => {
            // console.log(data);
            if (data) {
              weatherStore.fetchWeather(data[0]?.name);
              setUserCity(data[0]?.name);
            }
          })
          .catch(err => console.error(err));

      fetchUserCity();
    }
  }, [userCoordinates.lat, userCoordinates.lon, weatherStore]);

  if (initialLoading)
    return <RotateLoader cssOverride={override} role="status" aria-label="Loading Spinner" color="#36ddbc" />;

  if (error) return <article className={styles.error}>Sorry, {error?.message}. Try again!</article>;

  return (
    <section>
      <div>
        <p className={styles.emojiOne}> â›ˆ </p>

        <h2 className={styles.heading}>Find your current weather forecast</h2>
      </div>

      <div className={styles.input}>
        <Input
          label="Enter your city"
          placeholder={userCity}
          value={searchText}
          onChange={e => setSearchText(e.target.value)}
        />
      </div>

      <div className={styles.button}>
        <Button
          label="Get Forecast"
          onClick={() => {
            weatherStore.fetchWeather(searchText);

            setSearchQueries([...searchQueries, ...searchText.split("0")]);
          }}
        />
      </div>

      <h3>Your City: {userCity}</h3>

      <div className={styles.search}>
        {searchQueries.length > 0 && <h4 className={styles.searchHeading}>search history:</h4>}

        <ul>
          {searchQueries.map((item, i) => (
            <li key={`query${i}`} className={styles.queryList}>
              {item},
            </li>
          ))}
        </ul>
      </div>

      {weather ? (
        <article>
          <div>
            <p className={styles.paragraph}>
              <strong>
                <span className={styles.emojiTwo}> ðŸŒ¤ </span>
                <br />
                {weather && weather.name}
              </strong>
            </p>
          </div>

          <div className={styles.article}>
            <p>
              temp: {weather && weather.main.temp}, humidity: {weather && weather.main.humidity},
            </p>

            <ul>
              {weather &&
                weather.weather.map(item => (
                  <li key={item.id} className={styles.weatherList}>
                    {item.description}
                  </li>
                ))}
            </ul>
          </div>
        </article>
      ) : null}
    </section>
  );
});
export default Weather;
