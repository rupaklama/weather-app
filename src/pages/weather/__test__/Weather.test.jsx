/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable prettier/prettier */
/* eslint-disable no-undef */
import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { setupServer } from "msw/node";
import { rest } from "msw";

import Weather from "../Weather";
import { StoreContext, store } from "../../../stores/store";

const data = {
  coord: { lon: -74.0776, lat: 40.7282 },
  weather: [{ id: 800, main: "Clear", description: "clear sky", icon: "01d" }],
  base: "stations",
  main: { temp: 42.67, feels_like: 34.95, temp_min: 39.07, temp_max: 45.14, pressure: 1013, humidity: 49 },
  visibility: 10000,
  wind: { speed: 16.11, deg: 280 },
  clouds: { all: 0 },
  dt: 1673299664,
  sys: { type: 2, id: 2008776, country: "US", sunrise: 1673266801, sunset: 1673300774 },
  timezone: -18000,
  id: 5099836,
  name: "Jersey City",
  cod: 200,
};

let counter = 0;

let query;

const server = setupServer(
  rest.get("https://api.openweathermap.org/data/2.5/weather", (req, res, ctx) => {
    counter += 1;

    query = req.url.searchParams.get("q");

    return res(ctx.status(200), ctx.json(data));
  })
);

beforeEach(() => {
  counter = 0;
  server.resetHandlers();
});

beforeAll(() => server.listen());

afterAll(() => server.close());

const setup = () => {
  render(
    <StoreContext.Provider value={store}>
      <Weather />
    </StoreContext.Provider>
  );
};

describe("The <Weather /> component", () => {
  describe("layout", () => {
    it("should render without any errors", () => {
      setup();

      const { asFragment } = render(<Weather />);
      expect(asFragment()).toMatchSnapshot();
    });

    it("should display a Sub Heading", () => {
      setup();

      const headingEl = screen.getByRole("heading", { name: /find your current weather forecast/i });
      expect(headingEl).toBeInTheDocument();
    });

    it("should have the Input initially empty", () => {
      setup();

      const searchInputEl = screen.getByRole("textbox", { name: /enter your city/i });
      expect(searchInputEl.value).toBe("");
    });

    it("should display a Button", () => {
      setup();

      const buttonEl = screen.getByRole("button", { name: /get forecast/i });
      expect(buttonEl).toBeInTheDocument();
    });
  });

  describe("interactions", () => {
    it("displays loader after api call", async () => {
      setup();

      const searchInputEl = screen.getByRole("textbox", { name: /enter your city/i });

      userEvent.type(searchInputEl, "jersey city");

      const buttonEl = screen.getByRole("button", { name: /get forecast/i });
      userEvent.click(buttonEl);

      expect(await screen.findByRole("status")).toBeInTheDocument();

      await screen.findByText(data.name);
    });

    it("sets user geo coords on mount", async () => {
      const mockGeolocation = {
        getCurrentPosition: jest.fn().mockImplementationOnce(success =>
          Promise.resolve(
            success({
              coords: {
                latitude: 40.7215682,
                longitude: -74.047455,
              },
            })
          )
        ),
      };

      global.navigator.geolocation = mockGeolocation;

      setup();

      expect(mockGeolocation.getCurrentPosition).toHaveBeenCalledTimes(2);

      await screen.findByRole("status");

      expect(await screen.findByRole("heading", { name: /your city/i })).toBeInTheDocument();
    });

    it("makes an api call with user input value on button submit", async () => {
      setup();

      const searchInputEl = screen.getByRole("textbox", { name: /enter your city/i });

      userEvent.type(searchInputEl, "jersey city");

      const buttonEl = screen.getByRole("button", { name: /get forecast/i });
      userEvent.click(buttonEl);

      await screen.findByRole("status");

      expect(await screen.findByText(data.name)).toBeInTheDocument();
      expect(counter).toBe(1);
    });

    it("displays search history after api call", async () => {
      setup();

      const searchInputEl = screen.getByRole("textbox", { name: /enter your city/i });

      userEvent.type(searchInputEl, "jersey city");

      const buttonEl = screen.getByRole("button", { name: /get forecast/i });
      userEvent.click(buttonEl);

      expect(await screen.findByRole("heading", { name: /search history/i })).toBeInTheDocument();
    });

    it("displays an error message on api call fail", async () => {
      server.use(
        rest.get("https://api.openweathermap.org/data/2.5/weather", (req, res, ctx) => {
          return res(ctx.status(400));
        })
      );

      setup();

      const searchInputEl = screen.getByRole("textbox", { name: /enter your city/i });

      userEvent.type(searchInputEl, "jersey city");

      const buttonEl = screen.getByRole("button", { name: /get forecast/i });
      userEvent.click(buttonEl);

      await screen.findByRole("status");

      expect(await screen.findByRole("article")).toBeInTheDocument();
    });
  });
});
