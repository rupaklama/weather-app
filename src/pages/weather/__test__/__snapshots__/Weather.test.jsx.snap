// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The <Weather /> component layout should render without any errors 1`] = `
<DocumentFragment>
  <section>
    <div>
      <p
        class="emojiOne"
      >
         â›ˆ 
      </p>
      <h2
        class="heading"
      >
        Find your current weather forecast
      </h2>
    </div>
    <div
      class="input"
    >
      <div>
        <label
          class="label"
          for="input-field"
        >
          Enter your city
        </label>
        <input
          class="input"
          id="input-field"
          placeholder=""
          value=""
        />
      </div>
    </div>
    <div
      class="button"
    >
      <button
        class="button"
        type="button"
      >
        Get Forecast
      </button>
    </div>
    <h3>
      Your City: 
    </h3>
    <div
      class="search"
    >
      <ul />
    </div>
  </section>
</DocumentFragment>
`;
